api/apps/geoprocessing/src/marxan-sandboxed-runner/adapters-blm/blm-partial-results.repository.ts

```diff
@@ -1,37 +1,61 @@
 import { Injectable } from '@nestjs/common';
-import { SandboxRunnerOutputHandler } from '../sandbox-runner-output-handler';
 import { Workspace } from '../ports/workspace';
-
-type SomePartialResults = {};
+import { existsSync, promises } from 'fs';
+import { MarxanOutputParserService } from '../adapters-shared/marxan-output-parser/marxan-output-parser.service';
+import { InjectRepository } from '@nestjs/typeorm';
+import { BlmPartialResultEntity } from './blm-partial-results.geo.entity';
+import { Repository } from 'typeorm';
+import { BestSolutionService } from '../adapters-shared/marxan-output-parser/best-solution.service';
+import { MarxanDirectory } from '../adapters-single/marxan-directory.service';
 
 @Injectable()
-export class BlmPartialResultsRepository
-  implements SandboxRunnerOutputHandler<SomePartialResults> {
+export class BlmPartialResultsRepository {
+  constructor(
+    private readonly marxanOutputParser: MarxanOutputParserService,
+    private readonly bestSolutionsService: BestSolutionService,
+    @InjectRepository(BlmPartialResultEntity)
+    private readonly repository: Repository<BlmPartialResultEntity>,
+    private readonly marxanDirectory: MarxanDirectory,
+  ) {}
+
   cancel(): Promise<void> {
     return Promise.resolve(undefined);
   }
 
-  /**
-   * save partial results (for given blm value)
-   * to some temporary table
-   *
-   * would be used within BlmRunnerService, passed down to RunnerService
-   */
-  async dump(
-    workspace: Workspace,
-    scenarioId: string,
-    stdOutput: string[],
-    stdErr: string[] | undefined,
-  ): Promise<SomePartialResults> {
-    return {};
-  }
-
-  dumpFailure(
+  async savePartialResult(
     workspace: Workspace,
     scenarioId: string,
-    stdOutput: string[],
-    stdError: string[],
+    blmValue: number,
   ): Promise<void> {
-    return Promise.resolve(undefined);
+    const { fullPath: fullOutputPath } = this.marxanDirectory.get(
+      'OUTPUTDIR',
+      workspace.workingDirectory,
+    );
+
+    if (!existsSync(fullOutputPath + `/output_sum.csv`)) {
+      throw new Error(`Output is missing from the marxan run.`);
+    }
+
+    const runsSummary = (
+      await promises.readFile(fullOutputPath + `/output_sum.csv`)
+    ).toString();
+
+    const runs = this.bestSolutionsService.map(
+      this.marxanOutputParser.parse(runsSummary),
+    );
+    const bestRun = runs.find((run) => run.best);
+
+    if (!bestRun) {
+      throw new Error(
+        `${scenarioId} scenario calibration doesn't have a best run for ${blmValue} BLM value`,
+      );
+    }
+
+    await this.repository.save({
+      blmValue,
+      boundaryLength: bestRun.connectivity,
+      scenarioId,
+      score: bestRun.score,
+    });
```

Comment:
Consider the scenario:

1. Run Calibration for [0.1-1.0] range (fails at the 6th blm value
2. Run Calibration for [2-10] range

Boom, partial results (and final!) contains 11 values - as `move` to `final` repository didn't have a chance to run and thus to clear partials.

Possible option is to put an arbitrary id of `calibration` that we values as inserted - and remove all values related to given scenario that are not using that arbitrary ID. (or just remove all partials

Reply:
But, when should be removed partial results with a different `calibrationId`? On each iteration? Maybe I should try catching a possible error and then removing all partial results for that specific scenario? It sound good to you?

Reply:
assuming we put [scenarioId, calibrationId], we can try to remove "`where: {scenarioId, calibrationId: Not(...)}` or something like that. Isn't perfect, so feel free to find out. Not sure if removing "partials" upon a fail is fine - might be this is expected for some cases and partial results are fine - we don't know yet ;)